File: backend/app/models/project.py
from datetime import datetime, timezone

from bson import ObjectId
from pydantic import BaseModel, Field


class Project(BaseModel):
    id: ObjectId | None = Field(default=None, alias="_id")
    org_id: ObjectId
    name: str
    description: str = ""
    base_url: str = ""
    global_setup: str = ""
    # 1.5 Webhook notification
    webhook_url: str = ""
    # 3.1 Change-aware test selection — optional git integration
    git_repo_url: str = ""
    git_token: str = ""  # stored as-is; rotate via project settings
    # 3.5 Release approval workflow
    release_approvers: list[ObjectId] = Field(default_factory=list)
    # 8.3 Configurable gate thresholds (None = use system defaults)
    inconclusive_gate_pct: float | None = None   # default 0.15 (15%)
    behavior_override_pct: float | None = None   # default 0.20 (20%)
    created_by: ObjectId
    created_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))
    updated_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))

    model_config = {
        "populate_by_name": True,
        "arbitrary_types_allowed": True,
        "json_encoders": {ObjectId: str},
    }

File: frontend/src/types/project.ts
export interface Project {
  id: string;
  org_id: string;
  name: string;
  description: string;
  base_url: string;
  global_setup: string;
  webhook_url: string;
  git_repo_url: string;
  release_approvers: string[];
  // 8.3 Configurable gate thresholds
  inconclusive_gate_pct: number | null;
  behavior_override_pct: number | null;
  created_by: string;
  created_at: string;
  updated_at: string;
  test_case_count: number;
}

export interface ProjectListResponse {
  items: Project[];
  total: number;
  page: number;
  page_size: number;
}

export interface CreateProjectRequest {
  name: string;
  description?: string;
  base_url?: string;
  global_setup?: string;
}

export interface UpdateProjectRequest {
  name?: string;
  description?: string;
  base_url?: string;
  global_setup?: string;
}
